src/rime_api.cc

```diff
@@ -68,7 +68,7 @@ RIME_API void RimeSetup(RimeTraits *traits) {
 
   setup_deployer(traits);
   if (PROVIDED(traits, app_name)) {
-    SetupLogging(traits->app_name);
+    SetupLogging(traits->app_name, traits->min_log_level, traits->log_dir);
```

Comment:
Hi, I think this line introduce an ABI breakage. See https://github.com/hchunhui/librime-lua/issues/25#issuecomment-567776565 .

Clients built with older librime will pass a `traits` without `min_log_level` and `log_dir` at runtime, so we may want to check `data_size` before we use the two new members.

Reply:
@hchunhui Thanks for the report. Fixed in https://github.com/rime/librime/commit/62bbeada3d4f84a08714d3bb83e174efd5eb1b8a

Reply:
I do not think that 62bbeada3d4f84a08714d3bb83e174efd5eb1b8a is going to work as expected.
`PROVIDED` is defined as:
```
// assuming member is a pointer in struct *p
#define PROVIDED(p, member) ((p) && RIME_STRUCT_HAS_MEMBER(*(p), (p)->member) && (p)->member)
```
`log_dir` is a pointer (`const char*`), so `PROVIDED(traits, log_dir)` might work, but `min_log_level` is of type `int` and value `0` is one of several valid values of this variable, so `PROVIDED(traits, min_log_level)` could return false negative.

Reply:
fixed. thanks.